challenges-41-45.yml

```diff
@@ -0,0 +1,200 @@
+turbo_stream:
+  question: |
+    What are Turbo Streams NOT used for?
+  explanation: |
+    Turbo Streams are used to update parts of the page without having to necesarily perform
+    a request.
+
+    They are very usefull for applications that have to update constantly to events that
+    aren't always generated by the user, like chat applications.
+  features:
+    multiple_choice:
+      choices:
+        a: |
+          They are used to update another part of a page when we submit
+          from a Turbo Frame
+        b: |
+          They are used to update a part of our page periodically without
+          having to perform continuous requests.
+        c: |
+          They are used to add and remove Turbo Frames to our page dynamically
+      correct: c
+
+turbo_stream_2:
+  question: |
+    I'm building a chat application and the main chat template is located at
+    `chat/index.stream.html.twig`.
+
+    How should I render this template from a Symfony controller?
+  explanation: |
+    The response object `TurboStreamResponse()` is what makes the response special
+    so that Turbo can handle it the right way in the Browser!
+  features:
+    multiple_choice:
+      choices:
+        a: |
+          ```php
+              return $this->render('chat/index.stream.html.twig', [
+                  ], new TurboStreamResponse());
+          ```
+        b: |
+          ```php
+              return $this->render('chat/index.stream.html.twig');
+          ```
+        c: |
+          ```php
+              return $this->render('chat/index.stream.html.twig', [
+                  ], new Response());
+          ```
+        d: |
+          ```php
+              return $this->render('chat/index.stream.html.twig', [
+                  new TurboStreamResponse(),
+              ]);
+          ```
+      correct: a
+
+turbo_stream_3:
+  question: |
+    In my chat application, I have a spot in the page template
+    where I need to render this main chat flow:
+
+    ```html
+    <div id="chat-messages">
+    </div>
+
+    <div id="chat-input">
+        <input type="text" id="chat-text">
+        <button id="chat-submit">Submit!</button>
+    </div>
+    ```
+
+    Now I need to have this chat stream template *inside* the `<div>` with
+    `id="chat-message"`.
+
+    How should I do it?
```

Comment:
I think the question can be more clear. If I'm understanding the situation correctly, how about:

> When the user clicks "Submit!", the form submits to a controller where we send a Stream
> update to Mercure. The goal of that stream template is to update the `chat-messages`
> element to contain the message.
> 
> What is the correct way to do that?

Reply:
Yes. We haven't introduced Mercure YET, so I'm getting a bit ahead of ourselves here. All I want to point out here is that when the user clicks the send button in the "form" section of the page, the chat messages will update in the "chat" section of the page. So may be to make it more clear, I could add the 2 frames?

Reply:
Here's the updated question:

---

In my chat application, I have a spot in the page template
    where I need to render this main chat flow:

    ```html
    <turbo-frame id="chat-messages">
        { % for message in chat.messages % }
        <p>
            <span class="user">{{ message.from }}</span>
    
            <span class="message">{{ message.text }}</span>
        </p>
        {% endfor %}
    </turbo-frame>

    <turbo-frame id="chat-input">
        <form method="POST" action="{{ path('app_stream_send_chat_message') }}">
            <input type="text" id="chat-text">
    
            <button name="chat-submit">Submit!</button>
        </div>
    <turbo-frame>
    ```
        
    When the user clicks "Submit!", the form submits to a controller where we send a Stream
    update back to the Browser. The goal of that stream template is to update the chat-messages
    element to contain the message.

    What is the correct way to do that?

Reply:
@weaverryan wdyt?

Reply:
Ship it!